Engine/DataFeeds/Enumerators/Factories/DataQueueFuturesChainUniverseDataCollectionEnumerator.cs

```diff
@@ -0,0 +1,125 @@
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Data;
+using QuantConnect.Data.UniverseSelection;
+using QuantConnect.Interfaces;
+using QuantConnect.Data.Market;
+using QuantConnect.Data.Auxiliary;
+using System.Collections;
+
+namespace QuantConnect.Lean.Engine.DataFeeds.Enumerators
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="ISubscriptionEnumeratorFactory"/> that reads
+    /// a live dataset provided by symbol universe provider into a <see cref="FuturesChainUniverseDataCollection"/> enumerator
+    /// </summary>
+    public class DataQueueFuturesChainUniverseDataCollectionEnumerator : IEnumerator<FuturesChainUniverseDataCollection>
+    {
+        private readonly List<BaseData> _symbolUniverse;
+        private readonly IEnumerator<BaseData> _underlying;
+        private readonly Symbol _symbol;
+        FuturesChainUniverseDataCollection _current;
+        bool _needNewCurrent;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataQueueFuturesChainUniverseDataCollectionEnumerator"/> class.
+        /// </summary>
+        /// <param name="symbol">Futures contract symbol</param>
+        /// <param name="underlying">Underlying enumerator</param>
+        /// <param name="symbolUniverse">Symbol universe of the data queue</param>
+        public DataQueueFuturesChainUniverseDataCollectionEnumerator(Symbol symbol, IEnumerator<BaseData> underlying, List<BaseData> symbolUniverse)
+        {
+            _symbolUniverse = symbolUniverse;
+            _underlying = underlying;
+            _needNewCurrent = true;
+            _symbol = symbol;
+
+            _current = new FuturesChainUniverseDataCollection { Symbol = _symbol, Data = symbolUniverse };
+        }
+
+        /// <summary>
+        /// Returns current futures chain enumerator position 
+        /// </summary>
+        public FuturesChainUniverseDataCollection Current
+        {
+            get
+            {
+                if (_underlying.Current == null)
+                {
+                    return null;
+                }
+
+                // if we need to update the enumerator, we update it wrapping around the underlying enumerator
+                if (_needNewCurrent)
+                {
+                    var current = (FuturesChainUniverseDataCollection)_current.Clone();
+
+                    current.Time = _underlying.Current.Time;
+                    current.EndTime = _underlying.Current.EndTime;
+
+                    _current = current;
+                    _needNewCurrent = false;
```

Comment:
Does this code need to be synchronized? We may need a lock here. Typically anytime we're checking a boolean value and then updating it we'll need to synchronize, unless the type is explicitly not expected to be used in a multithreading scenario (which may be the case here)